trin-core/src/portalnet/overlay.rs

```diff
@@ -175,12 +190,182 @@ impl<TContentKey: OverlayContentKey + Send, TMetric: Metric + Send>
 
     /// Process overlay uTP listener event
-    pub fn process_utp_event(&self, event: UtpListenerEvent) {
-        // TODO: Handle overlay uTP events
-        debug!(
-            "Got overlay uTP event: {event:?}, protocol id: {:?}",
-            self.protocol
-        );
+    pub fn process_utp_event(&self, event: UtpListenerEvent) -> anyhow::Result<()> {
+        match event {
+            UtpListenerEvent::ClosedStream(utp_payload, _, stream_id) => {
+                // Handle all closed uTP streams, currently we process only AcceptStream here.
+                // FindContent payload is processed explicitly when we send FindContent request.
+                if let AcceptStream(content_keys) = stream_id {
+                    self.process_accept_utp_payload(content_keys, utp_payload)?;
+                }
+            }
+            UtpListenerEvent::ResetStream(..) => {
+                // TODO: Handle reset streams
+            }
+        }
+        Ok(())
+    }
+
+    /// Process accepted uTP payload of the OFFER?ACCEPT stream
+    fn process_accept_utp_payload(
+        &self,
+        content_keys: Vec<RawContentKey>,
+        payload: UtpPayload,
+    ) -> anyhow::Result<()> {
+        // Process accepted uTP payload as SZZ decoded Container[payloads: List[ByteList, max_length=64]]
+        let content_values = ContentPayloadList::from_ssz_bytes(&payload)
+            .map_err(|err| anyhow!("Unable to decode uTP content payload: {err:?}"))?;
+        // Max content value length should be 64
+        if content_values.len() > 64 {
+            return Err(anyhow!("Received content values with len > 64."));
+        }
+        // Accepted content keys len should match content value len
+        if content_keys.len() != content_values.len() {
+            return Err(anyhow!(
+                "Content keys len doesn't match content values len."
+            ));
+        }
+
+        // Try to store accepted content into the database and propagate gossip the content
+        match self.protocol {
+            ProtocolId::History => {
+                // TODO: Verify history content data with Oracle
+
+                // Temporally store content key/value pairs to propagate here
+                let mut content_keys_values: Vec<(HistoryContentKey, ByteList)> = Vec::new();
+
+                for (content_key, content_value) in content_keys.into_iter().zip(content_values) {
+                    let history_key = HistoryContentKey::try_from(content_key);
+
+                    match history_key {
+                        Ok(key) => {
+                            // Store accepted content in DB
+                            self.store_accepted_content(&key, content_value.clone());
+                            content_keys_values.push((key, content_value))
+                        }
+                        Err(err) => {
+                            warn!("Unable to decode history content key: {err}");
+                            continue;
```

Comment:
If any one piece of data is invalid above, I expect we might drop the whole stream. So it stands out that this is weirdly inconsistent, that it would only individually drop malformed keys. Also, this is especially weird right? Shouldn't these be the keys that we already sent ourselves in an ACCEPT? It would be highly strange for them to be undecodable. I think this error message doesn't convey the strangeness.

Reply:
Yeah, good points, updating the warning and returning an error makes more sense here :+1: